src/with_alloc.rs

```diff
@@ -163,24 +207,53 @@ impl<T> RingBuffer<T> for AllocRingBuffer<T> {
     impl_ringbuffer!(readptr, writeptr);
 }
 
-impl<T> AllocRingBuffer<T> {
+impl<T, MODE: RingbufferMode> AllocRingBuffer<T, MODE> {
     /// Creates a `AllocRingBuffer` with a certain capacity. This capacity is fixed.
     /// for this ringbuffer to work, cap must be a power of two and greater than zero.
+    ///
+    /// # Safety
+    /// Only safe if the capacity is greater than zero, and a power of two.
+    /// Only if Mode == NonPowerOfTwo can the capacity be not a power of two, in which case this function is also safe.
     #[inline]
-    pub fn with_capacity_unchecked(cap: usize) -> Self {
+    unsafe fn with_capacity_unchecked(cap: usize) -> Self {
         Self {
             buf: Vec::with_capacity(cap),
             capacity: cap,
             readptr: 0,
             writeptr: 0,
+            mode: Default::default(),
         }
     }
+}
 
+impl<T> AllocRingBuffer<T, NonPowerOfTwo> {
+    /// Creates a `AllocRingBuffer` with a certain capacity. This capacity is fixed.
+    /// for this ringbuffer to work, and must not be zero.
+    ///
+    /// Note, that not using a power of two means some operations can't be optimized as well.
+    /// For example, bitwise ands might become modulos.
+    ///
+    /// For example, on push operations, benchmarks have shown that a ringbuffer with a power-of-two
+    /// capacity constructed with `with_capacity_non_power_of_two` (so which don't get the same optimization
+    /// as the ones constructed with `with_capacity`) can be up to 3x slower
+    ///
+    /// # Panics
+    /// if the capacity is zeroa
+    #[inline]
+    pub fn with_capacity_non_power_of_two(cap: usize) -> Self {
+        assert_ne!(cap, 0, "Capacity must be greater than 0");
+
+        // Safety: Mode is NonPowerOfTwo and we checked above that the capacity isn't zero
+        unsafe { Self::with_capacity_unchecked(cap) }
+    }
+}
+
+impl<T> AllocRingBuffer<T, PowerOfTwo> {
     /// Creates a `AllocRingBuffer` with a certain capacity. The actual capacity is the input to the
     /// function raised to the power of two (effectively the input is the log2 of the actual capacity)
     #[inline]
     pub fn with_capacity_power_of_2(cap_power_of_two: usize) -> Self {
-        Self::with_capacity_unchecked(1 << cap_power_of_two)
+        unsafe { Self::with_capacity_unchecked(1 << cap_power_of_two) }
```

Comment:
Missing safety comment, should just be that we generate a capacity to the power of two, therefore it is safe.